### PR TITLE
Automerge 2.0 blog post

### PR DESCRIPTION
**Pull request by alexjg**

### CODE REVIEW

File: blog/2023-01-17-automerge-2/index.md [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+## Local-first Software
+
+Since the rise of the cloud, developers have largely had to choose between building cloud software or traditional installed software. Although installed software has some reliability and performance benefits, cloud software has dominated the market. Cloud software makes implementing collaboration easy and includes ubiquitous access from any computing device. Unfortunately, the advantages of cloud software come at a significant price. Cloud software is fragile and prone to outages, rarely supports offline use, and can be prohibitively expensive to scale to large audiences.
+
+At Ink & Switch, we’ve been researching a model for developing software which we call [local-first software](https://www.inkandswitch.com/local-first/), with the goal of combining the best of both worlds: reliable, locally-executed software paired with scalable offline-friendly collaboration infrastructure.
+
+Over the last five years we’ve built a wide variety of prototypes, run research experiments, and worked with a number of production users to better understand the challenges of building local-first software. The central result of this work is the Automerge project.
```
Comment (issackelly) [inline blog/2023-01-17-automerge-2/index.md]: Interesting phrasing. I think of Automerge as "the current consumable output of the work at the lab" but maybe a bit early to consider it the central result.

File: blog/2023-01-17-automerge-2/index.md [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+## Automerge
+
+Automerge is a JSON-like data structure that preserves history and can reliably merge edits from any other source. The combination of these two capabilities allows developers to easily support the kind of real-time collaboration expected by users today with native offline support.
```
Comment (issackelly) [inline blog/2023-01-17-automerge-2/index.md]: JSON-lik is an important capability, but it also undersells how we use it. We've got lots of rich types built on bytearrays

File: blog/2023-01-17-automerge-2/index.md [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+## Reimplementing Automerge for Performance & Portability
+
+Earlier versions of Automerge were implemented in pure JavaScript. Our initial implementations were theoretically sound but much too slow and used too much memory for most production use cases.
+
+Furthermore, JavaScript support on mobile devices and embedded systems is limited. We wanted a fast and efficient version of Automerge that was available everywhere: in the browser, on any mobile device, and or even embedded devices like the ESP32.
+
+As a result, we decided to rewrite Automerge in Rust. In the browser, Rust can be compiled to WebAssembly. It can also be used as a native library through bindings for other programming languages. 
+
+The change should be mostly transparent for JavaScript users. We built a JS wrapper around the Rust core that provides a similar API to earlier versions of Automerge.
```
Comment (issackelly) [inline blog/2023-01-17-automerge-2/index.md]: From the chat's i've seen, it does seem like the type changes for things like Text might not be very transparent (and will be very common among existing users) but I think automerge is still very early in the adoption curve and most of us using it for production are pretty aware of these things.

File: blog/2023-01-17-automerge-2/index.md [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+To create Automerge 2.0, lab members Alex Good and Orion Henry teamed up with Andrew Jeffrey, Jason Kankiewicz (as well as several other open source collaborators) to polish and optimize the Rust implementation and develop the JS wrapper. The result is a codebase that is hundreds of times faster, radically more memory efficient, and both better tested and more reliable.
+
+:::caution
```
Comment (issackelly) [inline blog/2023-01-17-automerge-2/index.md]: non-sequitur, right? Is this an incredibly common request that we feel it needs to go here?

File: blog/2023-01-17-automerge-2/index.md [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+## Native Rich Text Support
+
+As with most CRDTs, Automerge originally focused on optimizing editing of plain-text. In the [Peritext paper](https://www.inkandswitch.com/peritext/) by Ink & Switch we discuss an algorithm for supporting rich text with good merging accuracy, and we are planning to integrate this algorithm into Automerge. Support for rich text will also make it easier to implement features like comments or cursor and selection sharing.
+
+## Improved Synchronization
+
+While we are very proud of the round-trip efficiency of Automerge synchronization on a per-document basis, we see big potential to improve sync performance of many documents at once. We also expect to provide other optimizations our users and sponsors have requested, such as more efficient first-document loading, run-length encoding of related changes, and enabling something akin to a Git “shallow clone” in the case of clients which don’t often require access to a document’s history.
+
+## Rust API Improvements
```
Comment (issackelly) [inline blog/2023-01-17-automerge-2/index.md]: Worth calling out autosurgeon here.

File: blog/2023-01-17-automerge-2/index.md [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+## Automerge-Repo
+
+We’ve worked hard to keep Automerge unopinionated and support a wide variety of deployment environments. We don’t require a particular network stack or storage system, and Automerge has been used successfully in both client-server web applications and peer-to-peer desktop software. Unfortunately, this technology-agnostic position has left a lot of the busy-work up to application developers, and asked them to learn a lot about distributed systems just to get started.
+
+Our upcoming library, Automerge-Repo, is a modular batteries-included approach to building web applications with Automerge. It works both in the browser (desktop and mobile) and in Node, and supports a variety of networking and storage adapters. There are even text editor bindings for Quill and Prosemirror as well as React Hooks to make it easy to get started quickly.
+
+We’re using it for some internal projects already, so expect more news soon, and feel free to get in touch if you’d like to be an early tester.
+
+# Conclusion
```
Comment (issackelly) [inline blog/2023-01-17-automerge-2/index.md:133]: Maybe move the note about mobile to postscript? The only other thing I can think of that's on my mind but not covered here is long-change-history and tomb stoning, which I haven't seen discussed in some time, but it's on my mind for a medium term "ready for production" issue.

File: blog/2023-01-17-automerge-2/index.md (hunk 2 above)
Comment (pvh) [reply blog/2023-01-17-automerge-2/index.md]: oh! i look forward to hearing more about this soon.

File: blog/2023-01-17-automerge-2/index.md (hunk 1 above)
Comment (pvh) [reply blog/2023-01-17-automerge-2/index.md]: ah, interesting interpretation. it's the central result of the local-first track, but only one of many ongoing research areas 'round the shop.

File: blog/2023-01-17-automerge-2/index.md (hunk 3 above)
Comment (pvh) [reply blog/2023-01-17-automerge-2/index.md]: I think this is now fixed thanks to some clever work by Alex... and it's not in 2.0/2.01 anyway :)

File: blog/2023-01-17-automerge-2/index.md [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+Earlier versions of Automerge were implemented in pure JavaScript. Our initial implementations were theoretically sound but much too slow and used too much memory for most production use cases.
+
+Furthermore, JavaScript support on mobile devices and embedded systems is limited. We wanted a fast and efficient version of Automerge that was available everywhere: in the browser, on any mobile device, and or even embedded devices like the ESP32.
+
+As a result, we decided to rewrite Automerge in Rust. In the browser, Rust can be compiled to WebAssembly. It can also be used as a native library through bindings for other programming languages. 
+
+The change should be mostly transparent for JavaScript users. We built a JS wrapper around the Rust core that provides a similar API to earlier versions of Automerge.
+
+To create Automerge 2.0, lab members Alex Good and Orion Henry teamed up with open source collaborators including Andrew Jeffrey, and Jason Kankiewicz to polish and optimize the Rust implementation and develop the JS wrapper. The result is a codebase that is hundreds of times faster than past releases, radically more memory efficient, and both better tested and more reliable.
```
Comment (jeffa5) [inline blog/2023-01-17-automerge-2/index.md]:
```suggestion
To create Automerge 2.0, lab members Alex Good and Orion Henry teamed up with open source collaborators including Andrew Jeffery, and Jason Kankiewicz to polish and optimize the Rust implementation and develop the JS wrapper. The result is a codebase that is hundreds of times faster than past releases, radically more memory efficient, and both better tested and more reliable.
```

File: blog/2023-01-17-automerge-2/index.md [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+
+Of course, even the most productive authors struggle to type an entire paper quite so quickly. Indeed, writing a paper can occur over months or even years, making both storage size on disk and load performance important as well.
+
+| Size on Disk | bytes      |
+| ------------ | ---------- |
+| plain text   | 107,121    |
+| automerge    | 129,062    |
+| naive        | 32,100,000 |
+
+The binary format works wonders in this example, encoding a full history for the document with only 30% overhead. That's less than a one additional byte per character! Compare that to a naive JSON encoding and which can be 300 bytes *per character*.
```
Comment (jeffa5) [inline blog/2023-01-17-automerge-2/index.md]:
```suggestion
The binary format works wonders in this example, encoding a full history for the document with only 30% overhead. That's less than a one additional byte per character! Compare that to a naive JSON encoding which can be 300 bytes *per character*.
```

File: blog/2023-01-17-automerge-2/index.md [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+
+This is useful for quite a few reasons: you can use it to implement real-time collaboration for an application without having to figure out tricky application-specific algorithms on the server. You can also use it to better support offline work. We think it has even more potential than just that.
+
+Since the rise of the cloud, developers have largely had to choose between building cloud software or traditional installed software. Although installed software has some reliability and performance benefits, cloud software has dominated the market. Cloud software makes implementing collaboration easy and includes ubiquitous access from any computing device. Unfortunately, the advantages of cloud software come at a high price. Cloud software is fragile and prone to outages, rarely supports offline use, and is literally expensive to scale to large audiences.
```
Comment (ept) [inline blog/2023-01-17-automerge-2/index.md]: Delete "literally"

File: blog/2023-01-17-automerge-2/index.md [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+
+With Automerge 2.0, we've brought together an efficient binary data format with fast updates, save, and load performance. Without getting too into the details, we accomplish this by storing data in order: a run of text inserted by an author will probably appear together as a single string in the binary file format. This makes it quick to materialize the state of the document.
+
+Let's take a look at some numbers. One of the most challenging benchmarks for CRDTs is realtime text collaboration. That's because a long editing session can result in hundreds of thousands of individual keystrokes to record and synchronize. Martin Kleppmann recorded the keystrokes that went into writing an academic paper and replaying that data has become a [popular benchmark](https://github.com/automerge/automerge-perf) for CRDTs.
+
+| Insert ~260k operations      | Timing (ms) | Memory (bytes) |
+| ---------------------------- | ----------- | -------------- |
+| Automerge 1.0                |  ~1,280,000 | ~3,000,000,000 |
```
Comment (ept) [inline blog/2023-01-17-automerge-2/index.md]: Should this be Automerge 0.14? Version 1.0 doesn't really exist

File: blog/2023-01-17-automerge-2/index.md [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+> -- Issac Kelly, CTO, Bowtie.
+
+## Performance: Speed, Memory and Disk
+
+Using a CRDT inherently comes with additional overhead: we have to track additional information in order to be able to correctly merge work from different sources. The goal of all CRDT authors is to find the right trade-offs between preserving useful history, reducing CPU overhead, and efficiently storing data in memory and on disk.
+
+With the Automerge project, our goal is to retain the full history of any document and allow an author to reconstruct any point in time on demand. As software developers we're accustomed to having this power: it's hard to imagine version control without history.
+
+With Automerge 2.0, we've brought together an efficient binary data format with fast updates, save, and load performance. Without getting too into the details, we accomplish this by storing data in order: a run of text inserted by an author will probably appear together as a single string in the binary file format. This makes it quick to materialize the state of the document.
```
Comment (ept) [inline blog/2023-01-17-automerge-2/index.md]: To someone not familar with CRDTs it may seem strange to even contemplate storing characters in any order other than document order. I'd be tempted to omit the second and third sentence of this paragraph.

File: blog/2023-01-17-automerge-2/index.md [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+| Yjs (what version?)          |       1,074 |     10,141,696 |
+| Automerge 2.0.2-unstable     |         661 |     22,953,984 |
+
+Of course, even the most productive authors struggle to type an entire paper quite so quickly. Indeed, writing a paper can occur over months or even years, making both storage size on disk and load performance important as well.
+
+| Size on Disk | bytes      |
+| ------------ | ---------- |
+| plain text   | 107,121    |
+| automerge    | 129,062    |
+| naive        | 32,100,000 |
```
Comment (ept) [inline blog/2023-01-17-automerge-2/index.md]: Rather than "naive", better to reference a particular Automerge version (say 0.14)

File: blog/2023-01-17-automerge-2/index.md [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+| Automerge 2.0.2-unstable |         438 |
+
+Loading the compressed document is fast as well, ensuring the best possible start-up time.
+
+While we are proud of these results, we will continue to invest in improved performance with each release as you can see with the preliminary numbers for the upcoming Automerge 2.0.2 release.
+
+A few notes about methodology before we move on. The particular implementation we used to run the benchmarks can be found [here](https://github.com/alexjg/automerge-perf-comparisons). These numbers were produced on Ryzen 9 7900X. The "timing" column is how long it takes to apply every single edit in the trace, whilst the "memory" common is the peak memory usage during this process.
+
+The improvements found in "2.0.2-unstable" mostly result from an upcoming improved API for text. Also note that the "automerge 1.0.1" here is actually the `automerge@1.0.1-preview-7` release. Automerge 1.0.1 was a significant rewrite from 1.0.0 and has a similar architecture to the Rust implementation. Improvements between 1.0.1 and 2.0.1 are a result of both optimization and adopting WebAssembly rather than an architectural change.
```
Comment (ept) [inline blog/2023-01-17-automerge-2/index.md]: "Automerge 1.0.1 was a significant rewrite from 1.0.0" should be "Automerge 1.0.1 was a significant rewrite from 0.14"

File: blog/2023-01-17-automerge-2/index.md [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+
+## Portability & Mobile Devices
+
+Because the core logic of Automerge is now built in Rust, we're able to port it more easily to a wide variety of environments and bind it to almost any language. We have users today who directly build on Automerge using the Rust APIs (and the supporting autosurgeon library). We also have a C-bindings API designed and contributed by `jkankiewicz`, and are excited to see the `automerge-go` implementation underway by Conrad Irwin.
```
Comment (ept) [inline blog/2023-01-17-automerge-2/index.md]: Links to autosurgeon, C bindings, automerge-go

File: blog/2023-01-17-automerge-2/index.md [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+
+In the future, we hope to provide bindings for other languages like Swift, Kotlin, and Python.
+
+One important note is that React-Native does not support WASM today. Developers building mobile applications will need to bind directly via C. If you're interested in either working on or sponsoring work on this problem, feel free to get in touch.
```
Comment (ept) [inline blog/2023-01-17-automerge-2/index.md:99]: I believe the official abbreviation is "Wasm", not "WASM". No hyphen in "React Native".

"bind directly via C" sounds strange. If the binding is via another language, it's not direct. Maybe rephrase as something like: "Developers building mobile applications with React Native will need to use Automerge as a native library via the C bindings."

File: blog/2023-01-17-automerge-2/index.md [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+
+# What’s Next
+
+With the release of Automerge 2.0 out the door, we will of course be listening closely to the community about their experience with the release, but in the months ahead, we expect to work on at least some of the following features:
+
+## Native Rich Text Support
+
+As with most CRDTs, Automerge originally focused on optimizing editing of plain-text. In the [Peritext paper](https://www.inkandswitch.com/peritext/) by Ink & Switch we discuss an algorithm for supporting rich text with good merging accuracy, and we are planning to integrate this algorithm into Automerge. Support for rich text will also make it easier to implement features like comments or cursor and selection sharing.
```
Comment (ept) [inline blog/2023-01-17-automerge-2/index.md]: No hyphen in "plain text"

File: blog/2023-01-17-automerge-2/index.md [hunk 17]
```diff
@@ -0,0 +1,145 @@
...
+
+## Automerge-Repo
+
+We’ve worked hard to keep Automerge unopinionated and support a wide variety of deployment environments. We don’t require a particular network stack or storage system, and Automerge has been used successfully in both client-server web applications and peer-to-peer desktop software. Unfortunately, this technology-agnostic position has left a lot of the busy-work up to application developers, and asked them to learn a lot about distributed systems just to get started.
```
Comment (ept) [inline blog/2023-01-17-automerge-2/index.md]: Not sure about the word "unopinionated" here. Maybe better to call it "platform-agnostic"?

Besides client-server web apps and p2p desktop software, also mention internals of cloud services (like Fly.io, for example) as another deployment context?

File: blog/2023-01-17-automerge-2/index.md [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+
+Part of the motivation for rewriting Automerge into Rust was to make it easier to bring first-class support to mobile platforms. We hope to embark on first-party wrappers soon.
+
+## Rust API Improvements
+
+The current Rust API is powerful but low-level, and as we’ve watched our early adopter friends work with the API it has become clear to us what abstractions we can provide to make their lives easier via tools like type annotation.
+
+## Improved Synchronization
+
+While we are very proud of the round-trip efficiency of Automerge synchronization on a per-document basis, we see big potential to improve sync performance of many documents at once. We also expect to provide other optimizations our users and sponsors have requested, such as more efficient first-document loading, run-length encoding of related changes, and enabling something akin to a Git “shallow clone” in the case of clients which don’t often require access to a document’s history.
```
Comment (ept) [inline blog/2023-01-17-automerge-2/index.md]: "very proud" is a bit strong for my taste, since there are several known efficiency problems with the sync protocol — even with a single document, the hash graph reconstruction is expensive and sending individual changes is verbose

"run-length encoding" seems a bit technical and is only one of several compression techniques we use. I'd prefer "better compression of related changes"

File: blog/2023-01-17-automerge-2/index.md [hunk 19]
```diff
@@ -0,0 +1,145 @@
...
+
+# Introducing Automerge 2.0
+
+Automerge 2.0 is here and ready for production. It’s our first supported release resulting from a ground-up rewrite. The result is a production-ready CRDT with huge improvements in performance and reliability. It's available in both Javascript and Rust, and includes Typescript types and C bindings for use in other ecosystems. Even better, Automerge 2.0 comes with improved documentation and, for the first time, support options for production users.
+
+## Automerge, CRDTs, and Local-first Software
+
+Before getting into the details of why we're excited about Automerge 2.0, let's take a bit of time to explain what Automerge is for anyone unfamiliar with the project.
+
+Automerge is a CRDT, or "conflict-free replicated data type", but if you're allergic to buzzwords you can just think of it as a version controlled data structure. Automerge lets you record changes made to data and then replay them in other places, reliably producing the same result in each. It supports JSON-like data, including arbitrarily nested maps and arrays, as well as some more advanced data types such as text and numeric counters.
```
Comment (ept) [inline blog/2023-01-17-automerge-2/index.md]: Link to https://crdt.tech/

Maybe mention briefly that independent modifications of the data structure can be merged automatically (hence the name of the library)

File: blog/2023-01-17-automerge-2/index.md [hunk 20]
```diff
@@ -0,0 +1,145 @@
+---
+slug: automerge-2
+title: Automerge 2.0
+authors: [pvh]
+tags: []
+---
+
+# Introducing Automerge 2.0
+
+Automerge 2.0 is here and ready for production. It’s our first supported release resulting from a ground-up rewrite. The result is a production-ready CRDT with huge improvements in performance and reliability. It's available in both Javascript and Rust, and includes Typescript types and C bindings for use in other ecosystems. Even better, Automerge 2.0 comes with improved documentation and, for the first time, support options for production users.
```
Comment (ept) [inline blog/2023-01-17-automerge-2/index.md]: capitalisation: "TypeScript"

**Review by ept — COMMENTED** [review]
Nice blog post! I've left a few copyediting notes.

File: blog/2023-01-17-automerge-2/index.md (hunk 11 above)
Comment (pvh) [reply blog/2023-01-17-automerge-2/index.md]: iterated, thanks!

File: blog/2023-01-17-automerge-2/index.md (hunk 10 above)
Comment (pvh) [reply blog/2023-01-17-automerge-2/index.md]: technically it is 1.0-preview-something

File: blog/2023-01-17-automerge-2/index.md (hunk 12 above)
Comment (pvh) [reply blog/2023-01-17-automerge-2/index.md]: naive is the benchmark's input file

File: blog/2023-01-17-automerge-2/index.md (hunk 18 above)
Comment (pvh) [reply blog/2023-01-17-automerge-2/index.md]: I said we're only "very proud" of the round trip efficiency but I'll soften the praise.

File: blog/2023-01-17-automerge-2/index.md (hunk 12 above)
Comment (ept) [reply blog/2023-01-17-automerge-2/index.md]: According to [these notes](https://github.com/automerge/automerge/pull/253), the Automerge 0.14 JSON encoding of the same editing trace is 146 MB (6 MB as json+gzip). Maybe that would be a better comparison point than the ad-hoc encoding used by the benchmark data?
